### ISSUE TITLE
LogFiles should be streamed from the disk

### ISSUE THREAD

**Issue by Kittyfisto** [Kittyfisto/Tailviewer#41]
Currently, log files are read completely into memory which requires about 2x their size. This isn't practical for files that are GB in size.

Therefore a solution shall be implement so that:
a) Memory consumption is fixed to a constant upper limit, (mostly) independent of the number of files that are opened
b) Files should be streamed into memory on-demand, but not held in memory all the time (unless the aforementioned buffer allows it)
c) The implementation must perform well with filtered and merged log files (both often violate the locality principle and therefore a simple buffer for a block in one source file won't do)

**Comment by Kittyfisto** [comment]
This feature does NOT need to be implemented for 1.0, **however** it is quite probably that certain backwards incompatible changes will have to be made to ILogFile and its implementations. These changes must be identified and at least partially implemented so that plugins targeting tailviewer 1.0 will not break once this feature arrives with tailviewer 1.x

**Comment by Kittyfisto** [comment]
There are three fundamentally different ways to implement this feature:

## V1 ##
```csharp
interface ILogFile
{
    /// Method is non blocking, however the task finishes only once the data has been fetched
    Task GetEntriesAsync(LogFileSection section, ILogEntries buffer, int destinationIndex, CancellationToken token);
}
```
*Pros*:
- Different clients with different requirements are able to work with interfaces suited to their use case
*Cons*:
- All existing methods (GetEntries, etc..) need to be overloaded since every ILogFile client **besides** the UI can happily work with the blocking versions
  

## V2 ##
```csharp
interface ILogFile
{
    /// Method is blocks until data has been fetched, the UI needs to call this method from background threads
    void GetEntries(LogFileSection section, ILogEntries buffer, int destinationIndex);
}
...
Task.Factory.Run(() => logFile.GetEntries(...);
```
*Pros*:
- No changes to existing interface necessary
*Cons*:
- UI has no way to cancel a request (for example because that portion is no longer in view) which will cause the log file to serve a lot of irrelevant requests until the one the client is interested in is served
- This **really** sucks
  

## V3 ##
```csharp
interface ILogFile
{
    /// Method is does not block if the data is not present and immediately returns false
    /// Subsequent calls will eventually return true and cause the buffer to be filled
    bool TryGetEntries(LogFileSection section, ILogEntries buffer, int destinationIndex);
}
```
*Pros*:
- UI can easily work with this version (better than with V1)
- The implementation can easily server requests in FILO order, causing the most recent request to be served first (which is the behaviour the UI needs)
*Cons*:
- All existing methods (GetEntries, etc..) need to be overloaded since every ILogFile client **besides** the UI can happily work with the blocking versions

**Comment by Kittyfisto** [comment]
The decision has been made to use a variant of **V3**:

Instead of having a boolean value which tells the client if all the data has been retrieved, `BufferedLogSource` adds an additional, optional column which is set to true when the data was retrieved and to false otherwise.

This way, the ILogSource interface isn't changed for just **one** implementation and no other implementation has to care about what this column means.

**Comment by Kittyfisto** [comment]
Streaming from disk functionally works, however as soon as filtering comes into play, it becomes way too slow to be a viable alternative for the user.

The next step is to first investigate and then optimize the access patterns with which Tailviewer hits the text file.
Currently, there are several ILogSource implementations which are poorly implemented and hit the underlying data *repeatedly* and this needs to stop.

FilteredLogSource is the worst offender from all which retrieves data from the underlying source on a *column by column* basis.  This worked well enough when the entire log file was backed by memory, but now it will cause repeated reads of the same portions of data over and over.
MultiLineLogSource doesn't fare much better, but this is partly because there's still #140 to be done which should alleviate that considerably.

**Comment by Kittyfisto** [comment]
`ILogSource` (previously ILogFile) has been modified to support a first incarnation of (no further interface changes which are publicly visible are necessary) => Removing the "Breaks plugin compatibility" label as it's no longer true.
